posts/2022/09-15-testing-async.md

```diff
@@ -0,0 +1,82 @@
+
+```elixir
+# Application
+
+if Mix.env() === :test do
+  def processes, do: []
+else
+  def processes, do: [{SomeProcess, self()}]
+end
+```
+
+and then start the process in your test case
+
+```elixir
+# your test file
+
+setup do
+  start_supervised!({SomeProcess, self()})
+end
+```
+
+## Pass in the caller to fix Ecto sandbox issues
+
+Database calls are typically run within a transaction in the test environment so that any changes made during the test can be rolled back at the end. By default, one transaction is taken out per test case and it is linked to the PID of the test. This is called `manual` mode and is used when the `use DataCase` macro is set to `async: true`. The alternative mode is called `shared` and is used when the macro is set to `async: false` or just omitted entirely. In shared mode all processes running at the same time share the same transaction. Shared mode and synchronous tests go together because async tests need individual transactions to make sure that they are not sharing state. We want test transactions to be separate.
+
+When running an async process in manual mode, any database calls it makes will error because there was no transaction taken out for that PID (only for the main test process). It would be easy to set `async: false` on the `use DataCase` macro to resolve the issue, but the trouble with setting these tests to synchronous is that it dramatically slows down the runtime of the test suite, and we just need to come back and fix it later anyways.
+
+The solution is to use the `caller` option of Repo. If we pass in the PID of the main test process it will use that PID when finding the transaction rather than its own.
+
+```elixir
+MyApp.Repo.all(from(u in User), caller: test_pid)
+```
+
+When starting up the async process in `start_supervised!/2`, pass in the PID of the test process and use that for all repo calls. You can see an example of this in the last section.
+
+## Send messages to synchronize
+
+When it takes time for an async process to finish its job, we have to wait for it to complete before asserting the result. The first instinct could be to use `Process.sleep/1` and then assert after that delay. This can work but also leaves us with slow tests (because they will generally be waiting longer than necessary) and flaky tests (because sometimes they won't wait long enough). This is exacerbated by the difference in performance between a developer's machine and the container running the test on a continuous integration service. A better solution is to send a message from the async process to the test process to let it know that it is ready for the assertion.
+
+Create a function that only runs in the test environment.
+
+```elixir
+  @doc """
+  Provides a synchronization point for testing GenServers
+
+  assert_receive(:sync, 500) in test file
+  """
+  if Mix.env() === :test do
+    def maybe_send_sync(pid), do: send(pid, :sync)
+  else
+    def maybe_send_sync(_pid), do: :sync
+  end
+```
+
```

Comment:
Where does this code live?

Also: might be worth showing where you use the message sending in a test.
I find the article super interesting, but I'm not sure if I would be readily able to test that way based on the description in this post.

Can you include a full example below the text? Where it is all put together?

So something like:
Filename:
```code```

filename:
```code```

If that makes sense?